### ISSUE TITLE
RCO

### ISSUE THREAD

**Issue by johnvtan** [mfekadu/rc#25]
# Description
Add support for new R2 operators to `rco`. Because this pass comes after `shrink`, the only operators that need to be added are `eq?`, `<`, `not`, and `if` statements.

# Notes & Questions

## Basic `if` statement
`if` statements should be arranged such that the `then` case only evaluates if the `condition` is true, and the `else` case only evaluates if the `condition` is false.
```racket
; input expr
(if (not #t) (+ 2 (- 1)) (+ 3 (- 2)))

; BAD output - thn and els are evaluated even if the condition says otherwise
(let ([cnd (not #t)])
   ; pretend these aren't complex
   (let ([thn (+ 2 (- 1))]))
     (let ([els (+ 3 (- 2))])
       (if cnd thn els))))

; good output - thn and els only evaluated if condition says so
(let ([cnd (not #t)])
   (if cnd
     (let ([t1 (- 1)]) (+ 2 t1))
     (let ([t2 (- 2)]) (+ 3 t2))))
```

## Are `if` statements complex?
Given an expression with a nested `if` statement, such as:
```racket
(+ 2 (if #t 1 2))
```
should the expression be considered complex and therefore require rearrangement? It seems to make sense that nested `if` statements should be considered complex, since the outer addition operation cannot be completed until the `if` statement is evaluated. If the point of `rco` is to make it easier for `ec` to translate from `R2` to `C1` by making the order of execution in `R2` explicitly sequential instead of nested (ie, nested expressions must be evaluated before the outer operations), then it follows that this `if` statement must be considered complex.

#### 1) Explicitly rearrange code without using a temp
The above expression would be rearranged to look like:
```racket
(if #t (+ 2 1) (+ 2 2))
```
This might run into problems with more complex expressions, like
```racket
; input
(+ (+ 1 2) (if #t (+ 1 1) (- 1))

; output ?
(let ([t1 (+ 1 2)])
  (if #t 
     (let ([t2 (+ 1 1)]) (+ t1 t2))
     (let ([t3 (- 1)]) (+ t1 t3)))
```

The output with more complex subexpressions seems difficult to implement. The `thn` and `els` cases could contain an arbitrary number of nested `let` statements, but we would still need to keep track of `t1` until it's used at the bottom.

#### 2) Assign nested `if` to a temp variable
The above expression would be rearranged to look like:
```racket
(+ 2 (if #t 1 2))
(let ([t1 (if #t 1 2)]) (+ 2 t1))
```

A more complex example might look like:
```racket
; input
(+ (+ 1 2) (if #t (+ 1 1) (- 1))
(let ([t1 (if #t (+ 1 1) (- 1))]) (+ 2 t1))
```
This seems more doable to implement, but the main question I have is:
- How will `explicate-control` handle having `if` statements nested in the `val` part of a `let` binding?

I think this will be okay if we do it right, since this is the same as conditionally assigning a variable. If we see an `if` statement in the `val` of the `let` binding, it could be translated into `C1` as:
```
; output of rco
(let ([t1 (if #t (+ 1 1) (- 1))]) (+ 2 t1))

; potential pseudo-C1 output of ec
; C1 grammar states that if statement conditions must be of the form (cmp arg arg)
; see pg 60
(main (if (eq? #t #t) (goto label1) (goto label2))
(label1 (seq (assign t1 (+ 1 1)) (goto label3))  ; thn case - t1 is assigned to (+ 1 1)
(label2 (seq (assign t1 (- 1)) (goto label3))  ; els case - t1 is assigned to (- 1)
(label3 (return (+ 2 t1))      ; last statement gets executed after assignment

; In maybe a clearer way in C
int t1;
if (true) {
  t1 = 1 + 1;
} else {
  t1 = -1;
}
return 2 + t1;
```

**Comment by mfekadu** [comment]
The PR #32 implements **_solution 2_**.... and considering that, here are some possible implementation of `explicate-control`

## 1 
follow the book in implement explicate-control-pred


## 2
ignore the book because wouldn't the general ```[`(,op ,args ...)]``` case safely handle the `cnd` clause of the if-statement? Then modify the `ec-assign` function's let-matching-arm to have the following functionality

**GIVEN** from RCO-exp
```racket
(let ([t1 (if #t 1 2)]) (+ 2 t1))
```

**EXPECT** from explicate-control
```racket
((start (if (eq? #t #t) (goto L1) (goto L2))) 
(L1 (seq (assign t1 1) (goto L3) ))
(L2 (seq (assign t1 2) (goto L3) ))
(L3 (return (+ 2 t1)))
```
* notice the goto L3 helps to reduce duplication of code

**Comment by mfekadu** [comment]
The commit eb745d6e73d2f823d88aaf6a4224318ef83b879c describes an error inside RCO

**Comment by mfekadu** [comment]
The error described in the commit (eb745d6) is fixed by merging the branch "r2-rco" (#32). 

@johnvtan Let's just go ahead and merge #32?

because when I merged r2-rco into r2-ec, the error disappeared and rather we see an issue with select-instructions (see error below)

If there are RCO implementation details that explicate-control cares about then let's take care of that within the r2-ec branch. 




**```> ./run programs/if-should-exit-42.rc```**
```
select-instructions: handle-tail: handle-stmt: handle-expr failed on expression
  context...:
   /Applications/Racket v7.3/collects/racket/private/more-scheme.rkt:261:28
   [repeats 1 more time]
   "/Users/mfekadu/GitHub/rc/compiler.rkt": [running body]
   temp37_0
   for-loop
   run-module-instance!125
   perform-require!78

  context...:
   /Applications/Racket v7.3/collects/racket/private/more-scheme.rkt:163:2: select-handler/no-breaks
   /Applications/Racket v7.3/collects/racket/private/more-scheme.rkt:261:28
   [repeats 1 more time]
   "/Users/mfekadu/GitHub/rc/compiler.rkt": [running body]
   temp37_0
   for-loop
   run-module-instance!125
   perform-require!78

  context...:
   /Applications/Racket v7.3/collects/racket/private/more-scheme.rkt:163:2: select-handler/no-breaks
   /Applications/Racket v7.3/collects/racket/private/more-scheme.rkt:261:28
   "/Users/mfekadu/GitHub/rc/compiler.rkt": [running body]
   temp37_0
   for-loop
   run-module-instance!125
   perform-require!78

  context...:
   /Applications/Racket v7.3/collects/racket/private/more-scheme.rkt:163:2: select-handler/no-breaks
   "/Users/mfekadu/GitHub/rc/compiler.rkt": [running body]
   temp37_0
   for-loop
   run-module-instance!125
   perform-require!78
Traceback (most recent call last):
  File "build.py", line 31, in <module>
    do_stuff()
  File "build.py", line 24, in do_stuff
    subprocess.run(compile_args, stdout=output, check=True)
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/subprocess.py", line 487, in run
    output=stdout, stderr=stderr)
subprocess.CalledProcessError: Command '['./compiler.rkt', 'programs/if-should-exit-42.rc']' returned non-zero exit status 1.
>
```

**Comment by mfekadu** [comment]
done :) 🔥
